### PR TITLE
sm_ipt: Provide zephyr porting layer

### PR DESCRIPTION
**Pull request by dchat-nordic**
Updated zephyr porting layer for
nrf_rpc.
Added zephyr layer for sm_ipt (shared memory
inter-processor transport).

Depends on: 
https://github.com/nrfconnect/sdk-nrfxlib/pull/524 
https://github.com/nrfconnect/sdk-nrfxlib/pull/525

Signed-off-by: Dominik Chat <dominik.chat@nordicsemi.no>

### CODE REVIEW

File: subsys/CMakeLists.txt [hunk 1]
```diff
@@ -49,3 +49,4 @@ add_subdirectory(debug)
 add_subdirectory(partition_manager)
 
 add_subdirectory_ifdef(CONFIG_NRF_RPC nrf_rpc)
+add_subdirectory(sm_ipt)
```
Comment (KAGA164) [inline subsys/CMakeLists.txt:52]: Should it be added always ?

File: subsys/sm_ipt/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+zephyr_include_directories(include)
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_SM_IPT sm_ipt_os.c)
```
Comment (KAGA164) [inline subsys/sm_ipt/CMakeLists.txt:11]: Maybe just:
```suggestion
zephyr_library_sources(sm_ipt_os.c)
```

Add conditionally add sm_ipt directory `CONFIG_SM_IPT`

File: subsys/sm_ipt/include/sm_ipt_os.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+typedef atomic_t sm_ipt_os_atomic_t;
+#define sm_ipt_os_atomic_or atomic_or
+#define sm_ipt_os_atomic_and atomic_and
+#define sm_ipt_os_atomic_get atomic_get
+
+typedef struct k_mutex sm_ipt_os_mutex_t;
+#define sm_ipt_os_mutex_init k_mutex_init
+#define sm_ipt_os_unlock k_mutex_unlock
+static inline void sm_ipt_os_lock(sm_ipt_os_mutex_t *mutex) {
```
Comment (KAGA164) [inline subsys/sm_ipt/include/sm_ipt_os.h:50]: Please add blank line above

File: subsys/sm_ipt/include/sm_ipt_os.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	k_mutex_lock(mutex, K_FOREVER);
+}
+
+typedef struct k_sem sm_ipt_os_sem_t;
+#define sm_ipt_os_give k_sem_give
+static inline void sm_ipt_os_sem_init(sm_ipt_os_sem_t *sem) {
```
Comment (KAGA164) [inline subsys/sm_ipt/include/sm_ipt_os.h:56]: Please add blank line above

File: subsys/sm_ipt/sm_ipt_os.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+void sm_ipt_os_signal_handler(struct sm_ipt_os_ctx *os_ctx, void (*handler)(struct sm_ipt_os_ctx *))
+{
+	os_ctx->signal_handler = handler;
+}
+
+static void ipm_callback(const struct device *ipmdev, void *user_data, uint32_t id,
+			 volatile void *data)
+{
+	struct sm_ipt_os_ctx *os_ctx;
```
Comment (KAGA164) [inline subsys/sm_ipt/sm_ipt_os.c:35]:
```suggestion
	struct sm_ipt_os_ctx *os_ctx = (struct sm_ipt_os_ctx *) user_data;
```
Add you can remove next line

File: subsys/sm_ipt/sm_ipt_os.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	os_ctx = (struct sm_ipt_os_ctx *) user_data;
+
+	if (os_ctx->signal_handler != NULL)
```
Comment (KAGA164) [inline subsys/sm_ipt/sm_ipt_os.c:38]:
```suggestion
	if (os_ctx->signal_handler)
```

File: subsys/sm_ipt/sm_ipt_os.c [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+		os_ctx->signal_handler(os_ctx);
+}
+
+int sm_ipt_os_init(struct sm_ipt_os_ctx *os_ctx)
+{
+	const struct device *ipm_rx_handle;
+	uint32_t size = (SHM_SIZE / 8) * 4;
```
Comment (KAGA164) [inline subsys/sm_ipt/sm_ipt_os.c:45]: You can use solution from Zephyr

File: subsys/nrf_rpc/include/nrf_rpc_log.h [hunk 8]
```diff
@@ -15,7 +15,17 @@
 #define _NRF_RPC_LOG_REGISTER1(module) \
 	_NRF_RPC_LOG_REGISTER2(module)
 
+#define _NRF_RPC_LOG_DECLARE2(module) \
```
Comment (KAGA164) [inline subsys/nrf_rpc/include/nrf_rpc_log.h:18]: Why we need those two levels of macro ?

**Review by KAGA164 — CHANGES_REQUESTED** [review]

File: subsys/nrf_rpc/include/nrf_rpc_log.h (hunk 8 above)
Comment (dchat-nordic) [reply subsys/nrf_rpc/include/nrf_rpc_log.h:18]: If I were to use only level 2, then "CONFIG_ ## module ## _LOG_LEVEL", would result in appending name of module macro, not its value. Level 1 makes sure to pass only value of level 1s define.

File: west.yml [hunk 9]
```diff
@@ -104,7 +104,7 @@ manifest:
     - name: nrfxlib
       repo-path: sdk-nrfxlib
       path: nrfxlib
-      revision: 59179ff80cee01c4feb8b747e0ff9caecb55c92d
+      revision: d74c71a0362baa906031c24275a6bab110ef31fc
```
Comment (doki-nordic) [inline west.yml:107]: More convenient is following syntax. If you use it, you don't need to update this line each time dependent PR changes:
```
revision: pull/524/head
```

When dependent PR goes into master, you can remove this change.

**Review by doki-nordic — CHANGES_REQUESTED** [review]

**Comment by DatGizmo** [comment]
With the new Integration Test Specification process we don't need the github labels.
I will remove them.

Right now this PR runs all Integration tests because it changes sdk-nrfxlib in west.yml and there is no test-spec.yml in that repo yet.
You can add changes to this PR [nrfconnect/sdk-nrfxlib#687](https://github.com/nrfconnect/sdk-nrfxlib/pull/687) to make the test spec here more targeted

To get the test spec feature working correctly on this PR, you need to rebase it to the latest main to get the [.github/test-spec.yml](https://github.com/nrfconnect/sdk-nrf/blob/main/.github/test-spec.yml) file

Then run CI here to update the test spec.

[Information about the test spec feature](https://projecttools.nordicsemi.no/confluence/display/NCS/Test+Spec)
[CI-Help for issues](https://ci-help.nordicsemi.no/)

**Pull request closed without merging**